src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java

```diff
@@ -0,0 +1,262 @@
+import org.openmrs.reference.page.AddOrEditAllergyPage;
+import org.openmrs.reference.page.AllergyPage;
+import org.openmrs.reference.page.AttachmentsPage;
+import org.openmrs.reference.page.ConditionPage;
+import org.openmrs.reference.page.ConditionsPage;
+import org.openmrs.reference.page.PatientVisitsDashboardPage;
+import org.openmrs.reference.page.RequestAppointmentPage;
+import org.openmrs.reference.page.VisitNotePage;
+import org.openmrs.uitestframework.test.TestData;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class ClinicalVisitSteps extends Steps {
+
+	private static final String DIAGNOSIS_PRIMARY = "Cancer";
+	private static final String DIAGNOSIS_SECONDARY = "Malaria";
+	private static final String APPOINTMENT_TYPE = "Surgery";
+	private static final String DRUG_NAME = "Penicillins";
+	private static final String REACTION = "Headache";
+	private static final String CONDITION = "Malaria";
+	private ActiveVisitsPage activeVisitsPage;
+	private VisitNotePage visitNotePage;
+	private AllergyPage allergyPage;
+	private AddOrEditAllergyPage addOrEditAllergyPage;
+	private ConditionPage conditionPage;
+	private ConditionsPage conditionsPage;
+	private AttachmentsPage attachmentsPage;
+	private RequestAppointmentPage requestAppointmentPage;
+	private TestData.PatientInfo testPatient;
+
+	@Before(RunTest.HOOK.SELENIUM_CLINICAL_VISIT)
+	public void visitHomePage() {
+		testPatient = createTestPatient();
+		initiateWithLogin();
+		new TestData.TestVisit(testPatient.uuid, TestData.getAVisitType(),
+				getLocationUuid(homePage)).create();
+	}
+
+	@After(RunTest.HOOK.SELENIUM_CLINICAL_VISIT)
+	public void destroy() {
+		deletePatient(testPatient);
+		quit();
+	}
+
+	@Given("a user clicks on active visits link from home page")
+	public void launchPatientDashboardActiveVisits() {
+		activeVisitsPage = homePage.goToActiveVisitsSearch();
+	}
+
+	@When("a user selects a patient from active patient list")
+	public void searchActivePatient() {
+		activeVisitsPage.search(testPatient.identifier);
+		dashboardPage = activeVisitsPage
+				.goToPatientDashboardOfLastActiveVisit();
+	}
+
+	@Then("the system loads Patient dashboard page")
+	public void launchPatientDashboardPage() {
+		assertNotNull(dashboardPage.getActiveVisitList());
+	}
+
+	// User story: Complete visit note
+	@When("a user clicks visit note link from the patient dashboard")
+	public void loadVisitNotePage() {
+		visitsDashboardPage = dashboardPage.goToRecentVisits();
+		visitNotePage = (VisitNotePage) dashboardPage.goToVisitNote()
+				.waitForPage();
+	}
+
+	@Then("the system loads visit note page")
+	public void systemLoadsVisitNotePage() {
+		assertTrue(textExists("Visit Note"));
+	}
+
+	@When("a user fills the visit note")
+	public void fillVisitNoteForm() {
+		visitNotePage.selectProviderAndLocation();
```

Comment:
Why dont you simply include them in one single step,  i have seen you have simply splitted the functionality of visitNote into two steps, any reason why splitting than using one step handling some amount step definitions.

Reply:
> i have seen you have simply splitted the functionality of visitNote into two steps

@sherrif10 could you help me through more light on this? I seem not to get the point well !

Reply:
From line 78 to line 109 , if possible you can refactor that once into one single step definition since the idea is  to loadVisitNotePage enter visitNote details/ save and go back to patientdashboard.

Reply:
BDD framework uses feature files to hold user stories. These user stories tell what the user does on a functionality and documents it well. By combining codes from line `78` to `109` would mean some user steps will be encapsulated in almost one. For example the system can not add a visit note until a user fills the note form and then clicks on save button for the note to be saved  into the system. I would ideally say we rather ensure user stories on a functionality are well captured in each scenario than encapsulating many steps into one.

Reply:
i agree, all we are writing with gherkin syntax are user stories, so there is no difference , though am trying to ephasise clean and short more readable step definitions.

Reply:
> though am trying to emphasis clean and short more readable step definitions.

If we combine  user stories  of: `a user fills the visit note`, `a user clicks on save visit note button` and `the system adds note into visit note table` into one, wont this be encapsulating many steps into one? 

And this is not what we really want But rather capture every step in each scenario than encapsulating many steps into one!